tactics/equality.ml

```diff
@@ -915,17 +915,23 @@ let build_selector env sigma dirn c ind special default =
   let ind, _ = check_privacy env indp in
   let typ = Retyping.get_type_of env sigma default in
   let (mib,mip) = lookup_mind_specif env ind in
-  let deparsign = make_arity_signature env sigma true indf in
-  let p = it_mkLambda_or_LetIn typ deparsign in
-  let cstrs = get_constructors env indf in
-  let build_branch i =
-    let endpt = if Int.equal i dirn then special else default in
-    let args = List.map (fun d -> map_rel_decl EConstr.of_constr d) cstrs.(i-1).cs_args in
-    it_mkLambda_or_LetIn endpt args in
-  let brl =
-    List.map build_branch(List.interval 1 (Array.length mip.mind_consnames)) in
-  let ci = make_case_info env ind RegularStyle in
-  sigma, mkCase (ci, p, c, Array.of_list brl)
+  match mib.mind_record with
+  | None | Some None ->
```

Comment:
There is a function named `make_case_or_project`. Is it badly named or should it be used here?

Reply:
I guess it could be used, but its main function is essentially to recompute the information we have at hand in the `build_selector` function (as witnessed, in the primitive projection case the code is trivial)... So it's not very useful.

Reply:
> So it's not very useful.

Readability and reuse of the code is always useful. If that code had been reusing some other code, this bug might not even have existed in the first place.

Reply:
For what it's worth, [BZ#4622](https://coq.inria.fr/bugs/show_bug.cgi?id=4622) (and probably similar bugs) was fixed in #175 by introducing this `make_case_or_project`.
If it makes sense to use this function, I agree that it would be better to use it.

Reply:
All right, I'll update the code to use this function then.